### PR TITLE
Add the Teams API

### LINKED ISSUE #3: Team API

### PR DESCRIPTION
**Pull request by LeSeulArtichaut**
This will need some documentation before being merged 😅
Closes #3

### CODE REVIEW

File: src/api/teams.rs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            team = team_slug.into(),
+        );
+        self.crab.delete(url, None::<&()>).await
+    }
+
+    pub fn children(&self, team_slug: impl Into<String>) -> children::ListChildTeamsBuilder {
+        children::ListChildTeamsBuilder::new(self, team_slug.into())
+    }
+
+    pub fn repositories(&self, team_slug: impl Into<String>) -> team_repos::TeamRepoHandler {
```
Comment (XAMPPRocky) [inline src/api/teams.rs:59]:
```suggestion
    pub fn repos(&self, team_slug: impl Into<String>) -> team_repos::TeamRepoHandler {
```

File: src/api/teams/create.rs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        self.privacy = Some(privacy.into());
+        self
+    }
+
+    #[deprecated = "This has been deprecated by GitHub."]
+    pub fn permission(mut self, permission: impl Into<params::teams::Permission>) -> Self {
+        self.permission = Some(permission.into());
+        self
+    }
+
```
Comment (XAMPPRocky) [inline src/api/teams/create.rs:56]: I would prefer not to include anything that has already been deprecated by GitHub.
```suggestion
```

File: src/api/teams/edit.rs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        self
+    }
+
+    pub fn privacy(mut self, privacy: impl Into<params::teams::Privacy>) -> Self {
+        self.privacy = Some(privacy.into());
+        self
+    }
+
+    #[deprecated = "This has been deprecated by GitHub."]
+    pub fn permission(mut self, permission: impl Into<params::teams::Permission>) -> Self {
```
Comment (XAMPPRocky) [inline src/api/teams/edit.rs:41]: Same as above

File: src/api/teams/edit.rs [hunk 4]
```diff
@@ -0,0 +1,59 @@
+use super::*;
+use crate::params;
+
+#[derive(serde::Serialize)]
+pub struct EditTeamBuilder<'octo, 'a> {
```
Comment (XAMPPRocky) [inline src/api/teams/edit.rs:5]: Edit and Create seem extremely similar, with the only difference I can see is the send method. If that's the case I would recommend merging these two structs, and changing `slug` to be `Option<String>` and then use edit if present, and else use create. If the structs are merged I would also move this into `src/api/teams.rs` directly.

File: src/api/teams.rs [hunk 5]
```diff
@@ -0,0 +1,62 @@
+//! The Teams API
+
+mod children;
+mod create;
+mod edit;
+mod list;
+mod team_repos;
```
Comment (XAMPPRocky) [inline src/api/teams.rs:7]: The structs inside this need to publicly re-exported to viewable in documentation. The modules should still be private though.

File: src/api/teams/edit.rs (hunk 4 above)
Comment (LeSeulArtichaut) [reply src/api/teams/edit.rs:5]: There are 2 parameters in `CreateTeamBuilder` that don't exist in `EditTeamBuilder`: `maintainers` and `repo_names`. Is there a way to merge the structs while staying type-safe?

**Comment by LeSeulArtichaut** [comment]
(Applied your review suggestions)

File: src/api/teams/edit.rs (hunk 4 above)
Comment (XAMPPRocky) [reply src/api/teams/edit.rs:5]: Okay, I missed that there were two 2 parameters that were different,  it is fine then that they remain separate.

**Comment by LeSeulArtichaut** [comment]
@XAMPPRocky Just added the documentation. I didn't touch the README nor the crate-level documentation yet though.

**Comment by LeSeulArtichaut** [comment]
@XAMPPRocky Applied your suggestions.

**Comment by XAMPPRocky** [comment]
Seems like CI is still failing, the builder methods need to accept `impl Into<Option<T>>`.

**Comment by LeSeulArtichaut** [comment]
@XAMPPRocky Pushed to fix the doctests, CI should now pass.

**Comment by LeSeulArtichaut** [comment]
I didn't squash because I know you use GitHub's squash merge, please tell me if you still need me to squash my commits.

**Comment by XAMPPRocky** [comment]
No you don't ever need to squash, I will nearly always squash unless there's good reason to have separate commits in the PR.

File: src/api/teams.rs [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    ///     .teams("owner")
+    ///     .children("parent-team")
+    ///     .per_page(5)
+    ///     .page(1u8)
+    ///     .send()
+    ///     .await?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn children(&self, team_slug: impl Into<String>) -> ListChildTeamsBuilder {
```
Comment (XAMPPRocky) [inline src/api/teams.rs:142]: The builder/API methods should include the verb, to distinguish it from handler methods like `repos`.
```suggestion
    ///     .list_children("parent-team")
    ///     .per_page(5)
    ///     .page(1u8)
    ///     .send()
    ///     .await?;
    /// # Ok(())
    /// # }
    /// ```
    pub fn list_children(&self, team_slug: impl Into<String>) -> ListChildTeamsBuilder {
```

File: src/api/teams/team_repos.rs [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    /// # async fn run() -> octocrab::Result<()> {
+    /// let manages_repo = octocrab::instance()
+    ///     .teams("owner")
+    ///     .repos("team")
+    ///     .manages("owner", "repo")
+    ///     .await?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub async fn manages(
```
Comment (XAMPPRocky) [inline src/api/teams/team_repos.rs:40]:
```suggestion
    ///     .check_manages("owner", "repo")
    ///     .await?;
    /// # Ok(())
    /// # }
    /// ```
    pub async fn check_manages(
```

File: src/api/teams/team_repos.rs [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    ) -> Result<ManagesRepo> {
+        let url = format!(
+            "/orgs/{org}/teams/{team}/repos/{owner}/{repo}",
+            org = self.org,
+            team = self.team,
+            owner = repo_owner.into(),
+            repo = repo_name.into(),
+        );
+        let res = self.crab._get(&url, None::<&()>).await?;
+        Ok(match res.status() {
```
Comment (XAMPPRocky) [inline src/api/teams/team_repos.rs:53]: I would remove the `ManagesRepo` type, and instead return the alternative JSON response that contains the repository with permissions https://developer.github.com/v3/teams/#check-if-a-team-manages-a-repository Then people will be able to check what permissions the team has and in the simple case you can write
```rust
if octocrab::instance().teams("owner").repos("team-name").check_manages("owner", "repo").is_ok() {
  // ...
}

File: src/api/teams/team_repos.rs [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+    ///     .await?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub async fn add_or_update(
+        &self,
+        repo_owner: impl Into<String>,
+        repo_name: impl Into<String>,
+        permission: impl Into<Option<params::teams::Permission>>,
+    ) -> Result<()> {
```
Comment (XAMPPRocky) [inline src/api/teams/team_repos.rs:78]: Does `()` work here? I would assume it would fail to deserialise an empty response, but I'm not entirely sure.

Comment (LeSeulArtichaut) [reply src/api/teams/team_repos.rs:78]: Mhh, thinking about it I also think it would fail.

Comment (LeSeulArtichaut) [reply src/api/teams/team_repos.rs:78]: The docs in Serde for serde::Deserializer says:
> **unit:** The type of `()` in Rust. It represents an anonymous value containing no data.

As GitHub responds with 204 No Content, I'd expect `Response::text()` in reqwest to return an empty string.

So I set up a Playground test program:
```rust
use serde_json; // 1.0.52

fn main() {
    let v = serde_json::from_str::<()>("");
    println!("{:?}", v);
}
```

Which gave me:
> Err(Error("EOF while parsing a value", line: 1, column: 0))

Maybe a nice future enhancement to `serde_json`? :D
Anyway I'll fix this in the meantime.

**Comment by LeSeulArtichaut** [comment]
@XAMPPRocky I think I addressed all of your review comments

File: src/api/teams.rs [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+mod edit;
+mod list;
+mod team_repos;
+
+pub use self::{
+    children::ListChildTeamsBuilder,
+    create::CreateTeamBuilder,
+    edit::EditTeamBuilder,
+    list::ListTeamsBuilder,
+    team_repos::{ManagesRepo, TeamRepoHandler},
```
Comment (XAMPPRocky) [inline src/api/teams.rs:14]: This fixes CI.
```suggestion
    team_repos::TeamRepoHandler,
```

File: src/api/teams/team_repos.rs [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+        let url = format!(
+            "/orgs/{org}/teams/{team}/repos/{owner}/{repo}",
+            org = self.org,
+            team = self.team,
+            owner = repo_owner.into(),
+            repo = repo_name.into(),
+        );
+        Octocrab::map_github_error(self.crab._put(&url, permission.into().as_ref()).await?)
+            .await
+            .map(|_| ())
```
Comment (XAMPPRocky) [inline src/api/teams/team_repos.rs:83]: Very minor nit but I prefer using drop to map into a unit type.
```suggestion
            .map(drop)
```

File: src/api/teams.rs [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub async fn delete(&self, team_slug: impl Into<String>) -> Result<()> {
+        let url = format!(
+            "/orgs/{org}/teams/{team}",
+            org = self.owner,
+            team = team_slug.into(),
+        );
+        self.crab.delete(url, None::<&()>).await
```
Comment (XAMPPRocky) [inline src/api/teams.rs:125]:
```suggestion
        Octocrab::map_github_error(self.crab._delete(url, None::<&()>).await?).await.map(drop)
```

File: src/api/teams.rs [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub async fn delete(&self, team_slug: impl Into<String>) -> Result<()> {
+        let url = format!(
+            "/orgs/{org}/teams/{team}",
+            org = self.owner,
+            team = team_slug.into(),
+        );
+        Octocrab::map_github_error(self.crab._delete(url, None::<&()>).await?).await
```
Comment (XAMPPRocky) [inline src/api/teams.rs:125]:
```suggestion
        Octocrab::map_github_error(self.crab._delete(url, None::<&()>).await?).await.map(drop)
```

**Comment by LeSeulArtichaut** [comment]
CI now passes.

**Comment by XAMPPRocky** [comment]
Thank you again for your PR!

**Pull request merged**